File: src/sieve/c_soa/main_soa.c
// SPDX-FileCopyrightText: 2026 TSUKUMO Akito <tsukumoakito99@duck.com>
// SPDX-License-Identifier: MIT

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[]) {
  size_t n = 10000000;
  if (argc > 1) {
    n = atoll(argv[1]);
  }

  uint8_t *is_prime = (uint8_t *)malloc(n + 1);
  uint32_t *values = (uint32_t *)malloc((n + 1) * sizeof(uint32_t));

  if (!is_prime || !values) {
    return 1;
  }

  memset(is_prime, 1, n + 1);
  if (n >= 0)
    is_prime[0] = 0;
  if (n >= 1)
    is_prime[1] = 0;

  for (size_t i = 0; i <= n; i++) {
    values[i] = (uint32_t)i;
  }

  for (size_t p = 2; p * p <= n; p++) {
    if (is_prime[p]) {
      for (size_t i = p * p; i <= n; i += p) {
        is_prime[i] = 0;
      }
    }
  }

  uint64_t count = 0;
  for (size_t i = 0; i <= n; i++) {
    if (is_prime[i])
      count++;
  }

  printf("Checksum: %llu\n", (unsigned long long)count);
  fprintf(stderr, "Limit: %zu\n", n);

  free(is_prime);
  free(values);
  return 0;
}

File: src/sieve/zig_soa/main_soa.zig
// SPDX-FileCopyrightText: 2026 TSUKUMO Akito <tsukumoakito99@duck.com>
// SPDX-License-Identifier: MIT

const std = @import("std");
const Io = std.Io;

const NumberInfo = struct {
    is_prime: u8,
};

pub fn main(init: std.process.Init) !void {
    const args_allocator = init.arena.allocator();
    const args = try init.minimal.args.toSlice(args_allocator);

    var n: usize = 10000000;
    if (args.len > 1) {
        n = std.fmt.parseInt(usize, args[1], 10) catch 10000000;
    }

    const allocator = std.heap.page_allocator;
    var list = std.MultiArrayList(NumberInfo).empty;
    try list.ensureTotalCapacity(allocator, n + 1);
    defer list.deinit(allocator);
    list.len = n + 1;

    const sieve = list.items(.is_prime);
    @memset(sieve, 1);

    if (n >= 0) sieve[0] = 0;
    if (n >= 1) sieve[1] = 0;

    var p: usize = 2;
    while (p * p <= n) : (p += 1) {
        if (sieve[p] == 1) {
            var i = p * p;
            while (i <= n) : (i += p) {
                sieve[i] = 0;
            }
        }
    }

    var count: u64 = 0;
    for (sieve) |val| {
        if (val == 1) count += 1;
    }

    const stdout = Io.File.stdout();
    const stderr = Io.File.stderr();
    var out_buf: [128]u8 = undefined;
    var out_f = Io.Writer.fixed(&out_buf);
    try out_f.print("Checksum: {d}\n", .{count});
    try stdout.writeStreamingAll(init.io, out_f.buffered());

    var err_buf: [128]u8 = undefined;
    var err_f = Io.Writer.fixed(&err_buf);
    try err_f.print("Limit: {d}\n", .{n});
    try stderr.writeStreamingAll(init.io, err_f.buffered());
}
